### ISSUE TITLE
Disk based evictions

### ISSUE THREAD

**Issue by vishh** [kubernetes/kubernetes#25239]
Local disk is shared amongst pods and system daemons on a best effort today. Since we do not support disk as a first class resource, there is no means to restrict disk usage of pods.
Often times, nodes run out of local disk, and kubelet needs to free up disk space. Kubelet can perform several actions to free up disk space - delete old containers, delete unused images, evict pods that use a lot of disk. 
This issue tracks the overall effort of freeing up disk space on demand by the Kubelet.

[ ] Proposal for disk eviction PR - https://github.com/kubernetes/kubernetes/pull/25237
[ ] Refactor image and container garbage collections such that they can be run on demand based on eviction thresholds 
[ ] Extract logs outside of containers and delete containers immediately - depends on https://github.com/kubernetes/kubernetes/issues/24702
[ ] Introduce disk eviction thresholds
[ ] Delete logs and images based on thresholds
[ ] Add support for pod evictions
[ ] Generate disk pressure condition in Kubelet
[ ] Update Scheduler to not schedule pods whenever disk pressure condition is set on a node
[ ] e2e tests

**Comment by derekwaynecarr** [comment]
@vishh - I should be able to assist in the bottom 3 parts of Stretch.

**Comment by derekwaynecarr** [comment]
@vishh - I moved this out of 1.3.

**Comment by derekwaynecarr** [comment]
(but am still working on PRs that will sit...)

**Comment by vishh** [comment]
@derekwaynecarr Should we work against a common branch?

**Comment by derekwaynecarr** [comment]
Yeah.  That sounds good to me.  I will send a branch today or tomorrow.

**Comment by derekwaynecarr** [comment]
https://github.com/kubernetes/kubernetes/pull/27199

would address the following items:

- [x] Introduce disk eviction thresholds
- [x] Add support for pod evictions
- [x] Generate disk pressure condition in Kubelet
- [x] Update Scheduler to not schedule pods whenever disk pressure condition is set on a node
